<div style="background: linear-gradient(to right, #dc007e 0%, #dc007e 50%, #322e76 50%, #322e76 100%);"><br></div>

<img src="Documents/pacifico_logo.png" width="150" align="right"/>

# Tutorial Values Pacífico API

***

En este tutorial se mostrará el uso de la interfaz en Python para la API de Pacifico, más específicamente el uso para la solicitud de valores. La interfaz de Python para la API está en un paquete estándar de Python (Python Package Index: https://pypi.org/), el cual se llama "pacifico" (https://pypi.org/project/pacifico/) y podemos conseguir mediante el siguiente comando pip:

In [1]:
!pip install pacifico

  Using cached pacifico-0.0.5-py3-none-any.whl (33 kB)


Es importante notar que el carácter "!" antes del comando es exclusivo para ejecutar comandos desde el ambiente Jupyter, en una consola estándar no es necesario incluirlo.

***

## Importación

El primer paso para el uso de la librería o paquete es la importación de este:

In [4]:
import pacifico

Esta librería incluye todas las funcionalidades disponibles de la API en forma de funciones de Python.

***

## Autentificación

Para autentificarse se debe usar el token provisto por Pacifico. Esto se puede hacer con el archivo entregado (**token.key**) o directamente con el token como valor, pasándolo baje el argumento "token" de las funciones de la librería.

**Para este tutorial usaremos este método de autentificacion, por lo que deberá reemplazar el valor de ejemplo en la siguiente celda por su token personal para la API.**

In [11]:
# token = 'fO7tP81xaT6QUSsC321402gTu5MYot0K3ehMvtlC'  # Token de ejemplo, REEMPLAZAR
token = 'fOHbC81xaT6QUSss9I3u02gTu5MYot0K3rjZbylC'

***

## Consultar la información

Para consultar la información debemos, en primer lugar, acceder a la función "request" de la librería.

Para comenzar consultaremos el valor de un instrumento especifico, para esto debemos ingresar el nombre de este bajo el argumento "ticker" de la función. Como ejemplo, buscaremos el bono 'BCP0600323' que corresponde a un bono del Banco Central de Chile ('BCP').

In [14]:
data = pacifico.request(token, ticker='BCP0600323')
data

Scenario Date Publication      Date Effective Country Market Group Family  \
0  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
1  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
2  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
3  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
4  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
5  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
6  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   

       Ticker       Value        Field Date Tenor Other  
0  BCP0600323    3.558782    Convexity                   
1  BCP0600323    1.471278     Duration                   
2  BCP0600323    0.023100        Yield                   
3  BCP0600323    1.056235  Clean Price                   
4  BCP0600323    1.082901        Price                   
5  BCP0600323  105.623483  Clean Quote                   
6  BCP0600323  108.290150        Quote

También podemos buscar datos de la familia a la que pertenece este bono.

In [18]:
data_family = pacifico.request(token, family='BCP')
data_family

Scenario Date Publication      Date Effective Country Market Group Family  \
0   Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
1   Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
2   Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
3   Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
4   Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
5   Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
6   Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
7   Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
8   Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
9   Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
10  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
11  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
12  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
13  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
14  Pacifico       2021-01-29 2021-01-29 20:00:26   Chile     RF    BT    BCP   
15  Pacifico       2021-01-29 2021-01-29 20:00:26   Chile     RF    BT    BCP   
16  Pacifico       2021-01-29 2021-01-29 20:00:26   Chile     RF    BT    BCP   
17  Pacifico       2021-01-29 2021-01-29 20:00:26   Chile     RF    BT    BCP   
18  Pacifico       2021-01-29 2021-01-29 20:00:26   Chile     RF    BT    BCP   
19  Pacifico       2021-01-29 2021-01-29 20:00:26   Chile     RF    BT    BCP   
20  Pacifico       2021-01-29 2021-01-29 20:00:26   Chile     RF    BT    BCP   

        Ticker       Value        Field Date Tenor Other  
0   BCP0600322  105.199496        Quote                   
1   BCP0600322  102.532829  Clean Quote                   
2   BCP0600322    1.051995        Price                   
3   BCP0600322    1.025328  Clean Price                   
4   BCP0600322    0.014200        Yield                   
5   BCP0600322    0.539294     Duration                   
6   BCP0600322    0.813668    Convexity                   
7   BCP0600323  108.290150        Quote                   
8   BCP0600323  105.623483  Clean Quote                   
9   BCP0600323    1.082901        Price                   
10  BCP0600323    1.056235  Clean Price                   
11  BCP0600323    0.023100        Yield                   
12  BCP0600323    1.471278     Duration                   
13  BCP0600323    3.558782    Convexity                   
14  BCP0600221  102.996536        Quote                   
15  BCP0600221  100.029869  Clean Quote                   
16  BCP0600221    1.029965        Price                   
17  BCP0600221    1.000299  Clean Price                   
18  BCP0600221    0.004100        Yield                   
19  BCP0600221    0.008219     Duration                   
20  BCP0600221    0.008219    Convexity

También podemos buscar datos del grupo de esta familia y bono.

In [19]:
data_group = pacifico.request(token, group='BT')
data_group

Scenario Date Publication      Date Effective Country Market Group  \
0      Pacifico       2021-04-21 2021-04-21 21:37:34   Chile     RF    BT   
1      Pacifico       2021-04-21 2021-04-21 21:37:34   Chile     RF    BT   
2      Pacifico       2021-04-21 2021-04-21 21:37:34   Chile     RF    BT   
3      Pacifico       2021-04-21 2021-04-21 21:37:34   Chile     RF    BT   
4      Pacifico       2021-04-21 2021-04-21 21:37:34   Chile     RF    BT   
...         ...              ...                 ...     ...    ...   ...   
27182  Pacifico       2021-07-26 2021-07-26 21:07:35   Chile     RF    BT   
27183  Pacifico       2021-07-26 2021-07-26 21:07:35   Chile     RF    BT   
27184  Pacifico       2021-07-26 2021-07-26 21:07:35   Chile     RF    BT   
27185  Pacifico       2021-07-26 2021-07-26 21:07:35   Chile     RF    BT   
27186  Pacifico       2021-07-26 2021-07-26 21:07:35   Chile     RF    BT   

      Family      Ticker        Value      Field Date Tenor Other  
0         BR  BRA1220421  8801.683804      Quote                   
1         BR  BRA1220421    88.016838      Price                   
2         BR  BRA1220421     0.002100      Yield                   
3         BR  BRA1220421     0.002740   Duration                   
4         BR  BRA1220421     0.002736  Convexity                   
...      ...         ...          ...        ...        ...   ...  
27182     BR  BR71270721  9357.783628      Quote                   
27183     BR  BR71270721    93.577836      Price                   
27184     BR  BR71270721     0.005400      Yield                   
27185     BR  BR71270721     0.002740   Duration                   
27186     BR  BR71270721     0.002718  Convexity                   

[27187 rows x 12 columns]

Tambien podemos buscar datos del mercado de este grupo, familia y bono.

In [20]:
data_market = pacifico.request(token, market='RF')
data_market

Scenario Date Publication      Date Effective Country Market Group  \
0       Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    LH   
1       Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    LH   
2       Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    LH   
3       Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    LH   
4       Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    LH   
...          ...              ...                 ...     ...    ...   ...   
138686  Pacifico       2021-07-26 2021-07-26 21:07:35   Chile     RF    IF   
138687  Pacifico       2021-07-26 2021-07-26 21:07:35   Chile     RF    IF   
138688  Pacifico       2021-07-26 2021-07-26 21:07:35   Chile     RF    IF   
138689  Pacifico       2021-07-26 2021-07-26 21:07:35   Chile     RF    IF   
138690  Pacifico       2021-07-26 2021-07-26 21:07:35   Chile     RF    IF   

       Family        Ticker        Value        Field Date Tenor Other  
0          LH    CHI034 *05  8482.893245  Clean Price                   
1          LH    CHI034 *05     0.001100        Yield                   
2          LH    CHI034 *05     1.762772     Duration                   
3          LH    CHI034 *05     5.872835    Convexity                   
4          LH    CHI034 *05    28.646715        Quote                   
...       ...           ...          ...          ...        ...   ...  
138686     IF  FNBCI-270721    99.997135        Quote                   
138687     IF  FNBCI-270721     0.999971        Price                   
138688     IF  FNBCI-270721     0.010300        Yield                   
138689     IF  FNBCI-270721     0.002778     Duration                   
138690     IF  FNBCI-270721     0.000015    Convexity                   

[138691 rows x 12 columns]

Al ejecutar la función para obtener datos, esta nos devuelve por defecto un Pandas Dataframe con la información consultada.

***

## Metadata

Para saber cuál es el universo existente de datos para solicitar, debemos solicitar la metadata. Esta nos mostrara los autores, mercados, familias, grupos y tickers disponibles en el mismo formato en el que recibimos datos anteriormente.

Para solicitar la metadata debemos usar la función "request" sin ningún otro argumento más que el "token" para autentificar.

In [21]:
metadata = pacifico.request(token)
metadata

Scenario Date Publication      Date Effective        Country  \
0             Pacifico       2021-08-11 2021-08-11 16:37:48          Chile   
1             Pacifico       2021-08-11 2021-08-11 16:37:48          Chile   
2             Pacifico       2021-08-11 2021-08-11 16:37:48          Chile   
3             Pacifico       2021-08-11 2021-08-11 16:37:48          Chile   
4             Pacifico       2021-08-11 2021-08-11 16:37:48          Chile   
...                ...              ...                 ...            ...   
58169           BanRep       2021-08-11 2021-08-11 16:37:48       Colombia   
58170           BanRep       2021-08-11 2021-08-11 16:37:48       Colombia   
58171           BanRep       2021-08-11 2021-08-11 16:37:48       Colombia   
58172           BanRep       2021-08-11 2021-08-11 16:37:48       Colombia   
58173  Financial Times       2021-08-11 2021-08-11 16:37:48  United States   

      Market            Group Family        Ticker Value     Field Date Tenor  \
0         RF               IF     IF  F*CHI-030321        Metadata              
1         RF               IF     IF  F*BCI-060821        Metadata              
2         RF               IF     IF  F*EST-220421        Metadata              
3         RF               IF     IF  FNBNS-290421        Metadata              
4         RF               IF     IF  FNCOR-010321        Metadata              
...      ...              ...    ...           ...   ...       ...        ...   
58169    ECO        Inflacion    UVR         IPCCO        Metadata              
58170    ECO        Inflacion    UVR           UVR        Metadata              
58171    ECO        Inflacion    UVR        vIPCCO        Metadata              
58172    ECO  Tipos de Cambio    TRM           TRM        Metadata              
58173     RV               AC     CF  US74440Y1082        Metadata              

      Other  
0            
1            
2            
3            
4            
...     ...  
58169        
58170        
58171        
58172        
58173        

[58174 rows x 12 columns]

También podemos filtrar sobre la metadata, para esto usaremos los mismos parámetros de la función "request", pero en algún parámetro debemos pasar como valor el string "metadata" (en este ejemplo usamos el parámetro "ticker"). Por ejemplo, consultaremos la metadata exclusivamente para la los bonos del Banco Central de Chile ('BCP').

In [22]:
metadata_bcp = pacifico.request(token, family='BCP', ticker='metadata')
metadata_bcp

Scenario Date Publication      Date Effective Country Market Group Family  \
0  Pacifico       2021-08-11 2021-08-11 16:55:20   Chile     RF    BT    BCP   
1  Pacifico       2021-08-11 2021-08-11 16:55:20   Chile     RF    BT    BCP   
2  Pacifico       2021-08-11 2021-08-11 16:55:20   Chile     RF    BT    BCP   

       Ticker Value     Field Date Tenor Other  
0  BCP0600221        Metadata                   
1  BCP0600322        Metadata                   
2  BCP0600323        Metadata

De esta forma sabemos que estos son los tickers existentes en la API para los bonos del Banco Central de Chile ('BCP').

***

## Autores

Como se puede observar en la seccion de la metadata, existen varios autores (columna "Scenario") de datos dentro de la API, por lo que podemos especificar el autor o fuente de los datos al solicitarlos. Para hacer esto debemos especificar el autor bajo el argumento "author".

Como ejemplo pediremos datos del Banco Central de Chile (BCCh).

In [31]:
data_bcch = pacifico.request(token, author='bcch')
data_bcch

Scenario Date Publication Date Effective Country Market  \
0        BCCh       2020-11-01     2020-11-01   Chile    ECO   
1        BCCh       2020-11-01     2020-11-01   Chile    ECO   
2        BCCh       2020-11-01     2020-11-01   Chile    ECO   
3        BCCh       2020-11-01     2020-11-01   Chile    ECO   
4        BCCh       2020-11-01     2020-11-01   Chile    ECO   
...       ...              ...            ...     ...    ...   
6946     BCCh       2021-08-11     2021-08-11   Chile    ECO   
6947     BCCh       2021-08-11     2021-08-11   Chile    ECO   
6948     BCCh       2021-08-11     2021-08-11   Chile    ECO   
6949     BCCh       2021-08-11     2021-08-11   Chile    ECO   
6950     BCCh       2021-08-11     2021-08-11   Chile    ECO   

                       Group Family                     Ticker          Value  \
0           Tasas de Interes   F022  F022.DOT.STO.HM01.Z.CLP.M   13344.945569   
1           Tasas de Interes   F022  F022.DAV.STO.HM01.Z.CLP.M     228.395297   
2           Tasas de Interes   F022   F022.LIFP.STO.Z.NO.CLP.M     566.303857   
3           Tasas de Interes   F022     F022.CIT.PRO.Z.Z.USD.M      48.204612   
4           Tasas de Interes   F022   F022.DIFP.STO.Z.TC.CLP.M  141674.790953   
...                      ...    ...                        ...            ...   
6946         Tipos de Cambio   F072         F072.UAH.USD.N.O.D      26.654800   
6947         Tipos de Cambio   F072         F072.UYU.USD.N.O.D      43.480200   
6948               Inflacion   F073           F073.UFF.PRE.Z.D   29781.590000   
6949  Economia Internacional   F019         F019.PPB.PRE.100.D       4.237800   
6950  Economia Internacional   F019          F019.PRI.TAS.79.D       3.250000   

      Field Date Tenor                                              Other  
0     Quote             Otros depositos a menos de 1 mes en moneda nac...  
1     Quote             Depositos de ahorro a la vista a menos de 1 me...  
2     Quote             Inversiones financieras en moneda nacional (Sa...  
3     Quote             Colocaciones sector interfinanciero promedios ...  
4     Quote             Inversiones financieras en moneda extranjera (...  
...     ...        ...                                                ...  
6946  Quote                           Hryvnia ucraniano por dolar de EEUU  
6947  Quote                               Peso uruguayo por dolar de EEUU  
6948  Quote                                        Unidad de fomento (UF)  
6949  Quote             Precios de productos basicos / Onza troy de Co...  
6950  Quote                          Tasa de interes diaria Prime - dolar  

[6951 rows x 12 columns]

***

## Consulta Historica

Para consultar datos históricos, la API sólo tomará en cuenta las consultas que especifiquen fechas si es que también especifican los tickers a consultar. En el caso que no sea así devolverá los datos consultados, pero de hoy o la última fecha disponible, ignorando la especificación histórica.

Para especificar las fechas, hay que introducir las fechas como un objeto "datetime" en la función "request".

Como ejemplo, buscaremos el precio de la acción del Banco de Chile para los últimos 7 días.

In [23]:
import datetime

fechaInicio = datetime.date.today() - datetime.timedelta(days=7)
fechaTermino = datetime.date.today()

data = pacifico.request(token, ticker='CHILE', dateStart=fechaInicio, dateEnd=fechaTermino)
data

Scenario Date Publication      Date Effective Country Market Group Family  \
0   Pacifico       2021-08-04 2021-08-04 17:01:24   Chile     RV    AC     AC   
1   Pacifico       2021-08-04 2021-08-04 21:07:35   Chile     RV    AC     AC   
2   Pacifico       2021-08-04 2021-08-04 21:56:53   Chile     RV    AC     AC   
3   Pacifico       2021-08-05 2021-08-05 22:58:26   Chile     RV    AC     AC   
4   Pacifico       2021-08-06 2021-08-06 17:01:24   Chile     RV    AC     AC   
5   Pacifico       2021-08-06 2021-08-06 21:27:34   Chile     RV    AC     AC   
6   Pacifico       2021-08-09 2021-08-09 17:35:06   Chile     RV    AC     AC   
7   Pacifico       2021-08-10 2021-08-10 15:02:00   Chile     RV    AC     AC   
8   Pacifico       2021-08-10 2021-08-10 15:02:00   Chile     RV    AC     AC   
9   Pacifico       2021-08-10 2021-08-10 17:01:24   Chile     RV    AC     AC   
10  Pacifico       2021-08-10 2021-08-10 21:31:24   Chile     RV    AC     AC   
11      SPCH       2021-08-04 2021-08-04 00:00:00   Chile     RV    AC     AC   
12      SPCH       2021-08-05 2021-08-05 00:00:00   Chile     RV    AC     AC   
13      SPCH       2021-08-06 2021-08-06 00:00:00   Chile     RV    AC     AC   
14      SPCH       2021-08-09 2021-08-09 00:00:00   Chile     RV    AC     AC   
15      SPCH       2021-08-10 2021-08-10 00:00:00   Chile     RV    AC     AC   

   Ticker     Value  Field Date Tenor Other  
0   CHILE  66.40000  Price                   
1   CHILE  66.50000  Price                   
2   CHILE  66.50000  Price                   
3   CHILE  67.95000  Price                   
4   CHILE  67.95000  Price                   
5   CHILE  68.95000  Price                   
6   CHILE  68.95000  Price                   
7   CHILE  67.50000  Price                   
8   CHILE  67.50000  Price                   
9   CHILE  67.50000  Price                   
10  CHILE  68.84000  Price                   
11  CHILE  66.69718  Quote              CLP  
12  CHILE  66.50821  Quote              CLP  
13  CHILE  67.10780  Quote              CLP  
14  CHILE  68.62582  Quote              CLP  
15  CHILE  67.79253  Quote              CLP

***

## Convertir a otros formatos

Por defecto las consultas a la API mediante la librería devuelven un Pandas DataFrame con la información consultada. Pero existen otros dos formatos en que la API puede devolver esta información, en un diccionario de Python o en un string en formato JSON.

La manera en que especificamos el formato de salida de datos es mediante el argumento "format". Para el cual podemos especificar los 3 formatos como "dataFrame", "dictionary" y "json".

Ejecutaremos la misma consulta inicial para el bono 'BCP0600323' en los 3 distintos formatos. 

Para el Pandas DataFrame (por defecto):

In [28]:
data = pacifico.request(token, ticker='BCP0600323', format='dataFrame')
data

Scenario Date Publication      Date Effective Country Market Group Family  \
0  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
1  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
2  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
3  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
4  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
5  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   
6  Pacifico       2021-08-11 2021-08-11 12:21:08   Chile     RF    BT    BCP   

       Ticker       Value        Field Date Tenor Other  
0  BCP0600323    3.558782    Convexity                   
1  BCP0600323    1.471278     Duration                   
2  BCP0600323    0.023100        Yield                   
3  BCP0600323    1.056235  Clean Price                   
4  BCP0600323    1.082901        Price                   
5  BCP0600323  105.623483  Clean Quote                   
6  BCP0600323  108.290150        Quote

Para un diccionario de Python:

In [29]:
data = pacifico.request(token, ticker='BCP0600323', format='dictionary')
print(data)

{'Pacifico': {'11/08/2021': {'11/08/2021 12:21:08': {'Chile': {'RF': {'BT': {'BCP': {'BCP0600323': [{'value': 3.558781690954163, 'field': 'Convexity'}, {'value': 1.471277529349133, 'field': 'Duration'}, {'value': 0.0231, 'field': 'Yield'}, {'value': 1.056234829722584, 'field': 'Clean Price'}, {'value': 1.08290149638925, 'field': 'Price'}, {'value': 105.62348297225837, 'field': 'Clean Quote'}, {'value': 108.29014963892504, 'field': 'Quote'}]}}}}}}}}


Para el string en formato JSON:

In [30]:
data = pacifico.request(token, ticker='BCP0600323', format='json')
print(data)

{"Pacifico": {"11/08/2021": {"11/08/2021 12:21:08": {"Chile": {"RF": {"BT": {"BCP": {"BCP0600323": [{"value": 3.558781690954163, "field": "Convexity"}, {"value": 1.471277529349133, "field": "Duration"}, {"value": 0.0231, "field": "Yield"}, {"value": 1.056234829722584, "field": "Clean Price"}, {"value": 1.08290149638925, "field": "Price"}, {"value": 105.62348297225837, "field": "Clean Quote"}, {"value": 108.29014963892504, "field": "Quote"}]}}}}}}}}


Así existe flexibilidad en la entrega de datos por parte de la librería. 

***

## Guardar

Finalmente existen dos formas en las cuales podemos guardar los datos pedidos en una consulta, estos son como un archivo de texto o un archivo CSV y cual sea depende del formato en que se pidieron los datos.

Para poder guardar los datos debemos especificar bajo el argumento "fileName" el nombre del archivo donde los almacenaremos.

Para que la consulta se grabe como archivo de texto debemos pedir los datos en formato JSON o diccionario de Python y se guardaran con estructura JSON. Por otro lado, para que la consulta se grave como un archivo CSV debemos pedir los datos en formato Pandas DataFrame.

Volveremos a realizar el ejemplo para el bono 'BCP0600323' y lo guardaremos de las dos formas.

Para un archivo de texto:

In [13]:
data = pacifico.request(token, ticker='BCP0600323', format='json', fileName='pacifico_data')

Para un archivo CSV:

In [14]:
data = pacifico.request(token, ticker='BCP0600323', format='dataFrame', fileName='pacifico_data')

Por defecto los archivos serán guardados donde se está ejecutando este notebook o la librería.

En este caso luego de ejecutar los comandos tendremos los archivos "pacifico_data.txt" y "pacifico_data.csv", los cuales contienen los mismos datos, pero en distintos formatos.


***

## Descripción de los datos

### Agrupaciones de datos

Los datos que entrega la API tienen una jerarquía definida por el formato del archivo de entrega. Esto nos permite administrar los datos ordenadamente. Esta jerarquía es la siguiente:

- País
    - Mercado
        - Grupo
            - Familia
                - Ticker

Así, cada Ticker pertenece a una Familia, cada Familia pertenece a un Grupo, cada Grupo pertenece a un Mercado y cada Mercado pertenece a un País.

###### La nomenclatura usada por la API corresponde a la usada por la CMF, para más información el lector se puede referir al sitio: http://www.svs.cl/institucional/seil/certificacion_inst.php

### Fechas

Existen distintos tipos de fechas que definen características cruciales de los datos. A continuación, una breve descripción de cada uno de ellos:

- **Fecha de Publicación y Fixing**: Define la fecha en que el dato fue publicado y el fixing de la publicación. Si la hora de fixing es distinta al fin del día de publicación, entonces la función devolvera la fecha de publicación y la hora del fixing en intervalos de media hora, de lo contrario devolvera sólo la fecha de publicación.


- **Fecha Efectiva**: Corresponde a la fecha efectiva para el valor de este dato, por lo general esta será igual a la fecha de publicación, pero no necesariamente. Por ejemplo, cuando un precio es públicado al día desdepués de ser transado o calculado.


- **Fecha de Tenor**: Esta fecha, para los casos en que corresponda, mostrará la fecha del tenor del dato en cuestión. Un buen ejemplo, es el caso de un factor de descuento, aquí es vital saber para qué fecha del tenor está calculado el factor de descuento, y por lo tanto se incluirá esta fecha.


- **Fecha de Streaming**: Esta fecha corresponde a la fecha y hora en que el dato fue consultado por el cliente.

***

## Formato de Entrega

#### JSON

De manera general, los datos consultados desde la API serán entregados en formato JSON, y en una estructura que permite ordenar los datos obtenidos en base a sus características. Esta estructura, almacena características que pueden compartir los datos en capas. Así, los datos quedan seccionados en categorías, haciendo el archivo de entrega más liviano y ordenado.

El orden de las características o capas y la estructura general de entrega de la API es la siguiente:

<body>
<code>{Escenario: 
    {Fecha de publicación: 
        {Fecha efectiva: 
            {País: 
                {Mercado: 
                    {Grupo: 
                        {Familia: 
                            {Ticker:
                                [{
                                    "value": Valor del dato, 
                                    "field": Tipo de Dato,
                                    "dateTenor": Fecha del tenor del dato (en el caso que corresponda),
                                    "other": Característica extra del dato (en el caso que corresponda)
                                }]
                            }
                        }
                    }
                }
            }
        }
    }
}
</code>
</body>

Para poder ver esta estructura ejemplificada, tomaremos el bono 'BCP0600323' que obtuvimos en la primera sección del tutorial:

<body>
<code>{"Pacifico": 
    {"11/08/2021": 
        {"11/08/2021 12:21:08": 
            {"Chile": 
                {"RF": 
                    {"BT": 
                        {"BCP": 
                            {"BCP0600323": 
                                [{
                                    "value": 3.558781690954163, 
                                    "field": "Convexity"
                                }, {
                                    "value": 1.471277529349133, 
                                    "field": "Duration"
                                }, {
                                    "value": 0.0231, 
                                    "field": "Yield"
                                }, {
                                    "value": 1.056234829722584, 
                                    "field": "Clean Price"
                                }, {
                                    "value": 1.08290149638925, 
                                    "field": "Price"
                                }, {
                                    "value": 105.62348297225837, 
                                    "field": "Clean Quote"
                                }, {
                                    "value": 108.29014963892504, 
                                    "field": "Quote"
                                }]
                            }
                        }
                    }
                }
            }
        }
    }
}
</code>
</body>

Aquí podemos ver que el valor obtenido cuenta con un escenario de pricing, lo que significa que corresponde a un precio calculado (también está la posibilidad de que corresponda a una predicción). En este caso, no se incluye el nombre del autor del escenario debido a que corresponde a BFI. También, el resultado tiene como fecha de publicación el 5 de Octubre del 2020; la fecha efectiva del precio también corresponde al 5 de Octubre del 2020 y además incluye la hora 19:15:00; el país corresponde a Chile; el mercado corresponde a la renta fija (RF), el grupo corresponde a los Bonos Tesorería (BT), la familia corresponde a los Bonos de Banco Central de Chile (BCP); el ticker corresponde al mnemotécnico del bono; finalmente en la lista de datos tenemos dos datos, en primer lugar el valor (bajo la llave "value") que corresponde a 102.92331175431464 y el tipo de dato (bajo la llave "field") que corresponde a un precio, y en segundo lugar el valor (bajo la llave "value") que corresponde a 0.0023 y el tipo de dato (bajo la llave "field") que corresponde a una tasa.

#### Pandas DataFrame

Particularmente para la librería de Python existe la opción de entrega de datos en formato Pandas DataFrame, en donde cada columna corresponde a cada una de las características de los datos están almacenadas en columnas bajo las mismos nombres o llaves que el formato JSON. De esta manera cada dato se convierte en una fila del DataFrame, entregando de manera limpia los datos en un formato tabla.

Para poder ver esta estructura ejemplificada, tomaremos nuevamente el bono 'BCP0600323' que obtuvimos en la primera sección del tutorial:

|  | Scenario | Date Publication | Date Effective | Country | Market | Group | Family | Ticker | Value | Field | Date  Tenor | Other |
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| 0 | Pacifico | 2021-08-11 | 2021-08-11 12:21:08 | Chile | RF | BT | BCP | BCP0600323 | 3.558782 | Convexity |		
| 1 | Pacifico | 2021-08-11 | 2021-08-11 12:21:08 | Chile | RF | BT | BCP | BCP0600323 | 1.471278 | Duration |		
| 2 | Pacifico | 2021-08-11 | 2021-08-11 12:21:08 | Chile | RF | BT | BCP | BCP0600323 | 0.023100 | Yield |		
| 3 | Pacifico | 2021-08-11 | 2021-08-11 12:21:08 | Chile | RF | BT | BCP | BCP0600323 | 1.056235 | Clean Price |		
| 4 | Pacifico | 2021-08-11 | 2021-08-11 12:21:08 | Chile | RF | BT | BCP | BCP0600323 | 1.082901 | Price |		
| 5 | Pacifico | 2021-08-11 | 2021-08-11 12:21:08 | Chile | RF | BT | BCP | BCP0600323 | 105.623483 | Clean Quote |		
| 6 | Pacifico | 2021-08-11 | 2021-08-11 12:21:08 | Chile | RF | BT | BCP | BCP0600323 | 108.290150 | Quote |		

***

## Argumentos

El proceso de consulta de datos tiene múltiples variables de especificación, a partir de las cuales podemos acotar nuestras consultas de manera eficiente, estos son precisamente los argumentos que le entregaremos a la API. A continuación, están estos argumentos y sus descripciones:

- **Ticker**: Corresponde al ticker o tickers sobre los cuales se quieren consultar datos (ej. 'CHILE' para la acción del Banco de Chile).


- **Familia**: Corresponde a la familia o familias sobre los cuales se quieren consultar datos (ej. 'BCP' para los Bonos de Banco Central de Chile).


- **Grupo**: Corresponde al grupo o grupos sobre los cuales se quieren consultar datos (ej. 'BT' para los Bonos de Tesorería).


- **Mercado**: Corresponde al marcado o mercados sobre los cuales se quieren consultar datos (ej. 'RF' para el mercado de Renta Fija).


- **País**: Corresponde al país o países sobre los cuales se quieren consultar datos (ej. 'Chile' para Chile).


- **Fecha de Inicio**: Corresponde a la fecha de inicio cuando se quieren consultar datos históricamente.


- **Fecha de Término**: Corresponde a la fecha de término cuando se quieren consultar datos históricamente.


- **Fixing**: Corresponde al fixing o fixings sobre los cuales se quieren consultar datos (ej. 'F_0000' para las 00:00 horas).


- **Tipo de Datos**: Corresponde al tipo de datos sobre los cuales se quieren consultar datos (ej. 'Price' para los precios).


- **Tipo de Versión**: Corresponde al tipo de versión de los datos que se quieren consultar (ej. 'Pricing' cálculos sobre precios).


- **Autor**: Corresponde al autor de los datos que se quieren consultar (ej. 'BFI' para cálculos y publicaciones hechas por Burrus Financial Intelligence).


- **Versión**: Corresponde a la versión de los datos que se quieren consultar (ej. '' para la versión por defecto establecida por Burrus Financial Intelligence).


##### Valores por Defecto

Los valores por defecto de los argumentos son los siguientes:

- Ticker: ''.
- Familia: ''.
- Grupo: ''.
- Mercado: ''.
- País: ''.
- Fecha de Inicio: Fecha de hoy.
- Fecha de Término: Fecha de hoy.
- Fixing: 'EOD'.
- Tipo de Datos: 'Unspecified'.
- Tipo de Versión: 'Pricing'.
- Autor: 'Pacifico'.
- Versión: ''.

Estos son los valores por defecto debido a que al consultar datos con estos argumentos obtendremos todos los datos de hoy o de la última fecha para la que hay datos disponibles.

##### Funcion request

Los argumentos de la librería respecto a consultas de valores de Python corresponden a los argumentos de la función "request".

Los nombres de los argumentos en la función de la librería de Python son:

- Ticker: 'ticker'.
- Familia: 'family'.
- Grupo: 'group'.
- Mercado: 'market'.
- País: 'country'.
- Fecha de Inicio: 'dateStart'.
- Fecha de Término: 'dateEnd'.
- Fixing: 'fixing'.
- Tipo de Datos: 'dataType'.
- Tipo de Versión: 'versionType'.
- Autor: 'author'.
- Versión: 'version'.

***

## Factores de Mercado

Existen mercados especiales dentro del universo de los mercados, los que contienen todos los tickers que almacenan algún tipo de factor. Estos se consideran en un nivel mas bajo de datos ya que gracias a estos otros valores son calculados.

Entre estos existe una distinción gravitante, que aquellos mercados, grupos o familias de factores que tienen un asterisco (*) en el nombre, son aquellos factores que están interpolados en tenores estándar, mientras que aquellos que no lo tengan son valores calculados directamente.

Los factores de mercado se almacenan con la siguiente estructura:

- Mercados:

| Mercado | Descripción |
| :-: | :-: |
| MF | Factores de Mercado |
| MF* | Factores de Mercado* (interpolados en tenores estándar) |

- Grupos:

| Mercado | Grupo | Descripción |
| :-: | :-: | :-: |
| MF | CF | Factores de crédito |
| MF | DF | Factores de descuento |
| MF | PF | Factores de paridad |
| MF* | CF* | Factores de crédito* (interpolados en tenores estándar) |
| MF* | DF* | Factores de descuento* (interpolados en tenores estándar) |

En el caso de las familias, los factores de crédito y paridad son demasiados, por lo que solo mostraremos algunos ejemplos.

- Familias (factores de descuento / factores de descuento*): 

| Mercado | Grupo | Familia | / | Mercado | Grupo | Familia | 
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | 
| MF | DF | CLP@Chile | / | MF* | DF* | CLP@Chile* |
| MF | DF | CLP@TAB30 | / | MF* | DF* | CLP@TAB30* | 
| MF | DF | CLP@TPM | / | MF* | DF* | CLP@TPM* |  
| MF | DF | EUR@E1M | / | MF* | DF* | EUR@E1M* | 
| MF | DF | EUR@E3M | / | MF* | DF* | EUR@E3M* | 
| MF | DF | EUR@EON | / | MF* | DF* | EUR@EON* | 
| MF | DF | EUR@ESTR | / | MF* | DF* | EUR@ESTR* | 
| MF | DF | EUR@USA | / | MF* | DF* | EUR@USA* | 
| MF | DF | IPSA@Chile | / | MF* | DF* | IPSA@Chile* | 
| MF | DF | IPSAN@Chile | / | MF* | DF* | IPSAN@Chile* | 
| MF | DF | UF@Chile | / | MF* | DF* | UF@Chile* | 
| MF | DF | UF@TABUF90 | / | MF* | DF* | UF@TABUF90* | 
| MF | DF | USD@Chile | / | MF* | DF* | USD@Chile* | 
| MF | DF | USD@FED | / | MF* | DF* | USD@FED* |
| MF | DF | USD@L1M | / | MF* | DF* | USD@L1M* | 
| MF | DF | USD@L1Y | / | MF* | DF* | USD@L1Y* | 
| MF | DF | USD@L3M | / | MF* | DF* | USD@L3M* | 
| MF | DF | USD@L6M | / | MF* | DF* | USD@L6M* | 
| MF | DF | USD@SOFR | / | MF* | DF* | USD@SOFR* | 
| MF | DF | USD@USA | / | MF* | DF* | USD@USA* | 

- Familias (factores de crédito / factores de crédito*):

Un ejemplo para los factores de crédito y crédito*, son las curvas "97004000-5@CLP" y "97004000-5@CLP$^*$", que corresponden a las curvas de crédito del Banco de Chile (rut 97.004.000-5) en CLP, en cálculos directos e interpolaciones en tenores estándar ($^*$) respectivamente.

- Familias (factores de paridad): 

Un ejemplo para los factores de paridad es la curva "USD-CLP", que corresponden a las curvas de paridad del tipo de cambio del dólar en CLP.


***